### PR TITLE
Replace simplex transform with one based on the ILR transform

### LINKED ISSUE #3170: Add inverse ILR simplex transform
Add inverse ILR simplex. 

@WardBrian we can do this in 2 loops using the online softmax https://arxiv.org/abs/1805.02867. The first loop constructs the sum to zero, get the max value of that sum to zero vec, and returns the sum of exponentials with the max subtracted. The second loop does the safe exponential of the sum to zero vector with the max subtracted and divides by the sum of exponentials output from the first loop. The jacobian is output after.

```cpp
inline plain_type_t<Vec> simplex_ilr_constrain(const Vec& y, Lp& lp) {
  const auto N = y.size();

  plain_type_t<Vec> z = Eigen::VectorXd::Zero(N + 1);
  if (unlikely(N == 0)) {
    return z;
  }

  auto&& y_ref = to_ref(y);
  value_type_t<Vec> sum_w(0);

  // new   
  double d = 0;  // sum of exponentials
  double max_val = 0;
  double max_val_old = 0;

  for (int i = N; i > 0; --i) {
    double n = static_cast<double>(i);
    auto w = y_ref(i - 1) * inv_sqrt(n * (n + 1));
    sum_w += w;

    z.coeffRef(i - 1) += sum_w;
    z.coeffRef(i) -= w * n;
    
    // new
    max_val = max(max_val_old, z.coeff(i));
    d = d * exp(max_val_old - max_val) + exp(z.coeff(i) - max_val);
    max_val_old = max_val;
  }

  // new loop
  for (int i = 0; i < N; ++i) {
   z.coeffRef(i) = exp(z.coeff(i) - max_val) / d;
  }

  lp += -N * log(d) + 0.5 * log(N);

  return z;
 }
```

### PR DESCRIPTION
**Pull request by WardBrian**
## Summary

Closes #3170.

This updates the simplex transform and the stochastic matrices to use the ILR transform rather than stick breaking. 
The stochastic matrices delegate more to the simplex functions for now, they could be rewritten for efficiency later.

Q: This completely replaces the existing functions, but we may want to preserve them under a different name. If we're fine replacing them, this also closes #3168 

## Tests

Existing simplex tests pass, and I added one more to test some edge behavior

## Release notes

The simplex transform is now defined in terms of the isometric log ratio transform instead of stick breaking.

## Checklist

- [x] Copyright holder: Simons Foundation

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Coding-Style-and-Idioms) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by WardBrian** [comment]
A quick validation:

Using the `multi-logit-normal` from https://github.com/bob-carpenter/transforms

<details><summary>Code and data</summary>
<p>

```stan
functions {
  real multi_logit_normal_cholesky_lpdf(vector theta, vector mu,
                                        matrix L_Sigma) {
    int N = rows(theta);
    vector[N] log_theta = log(theta);
    return multi_normal_cholesky_lpdf(log_theta[1 : N - 1] - log_theta[N] | mu, L_Sigma)
           - sum(log_theta);
  }
}
data {
  int<lower=1> N;
  vector[N - 1] mu;
  matrix[N - 1, N - 1] L_Sigma;
}
parameters {
  simplex[N] x;
}
model {
  target += multi_logit_normal_cholesky_lpdf(x | mu, L_Sigma);
}
```

```json
{
  "N": 5,
  "mu": [-0.0104, -0.006, -0.0062, -0.0081],
  "L_Sigma": [
    [20.8371, 0.0, 0.0, 0.0],
    [65.0306, 58.6011, 0.0, 0.0],
    [-63.5698, 25.1499, 30.1392, 0.0],
    [32.0142, 22.2009, 69.8059, 21.1121]
  ]
}

```

</p>
</details> 

Here are the stansummarys of the current transform (`sb`):

```
$ ./bin/stansummary sb_1.csv sb_2.csv sb_3.csv sb_4.csv

Inference for Stan model: multi_normal_logit_model
4 chains: each with iter=1000; warmup=1000; thin=1; 1000 iterations saved.

Warmup took (0.18, 0.13, 0.14, 0.12) seconds, 0.57 seconds total
Sampling took (0.086, 0.12, 0.080, 0.10) seconds, 0.39 seconds total

                 Mean   MCSE  StdDev      MAD       5%      50%    95%  ESS_bulk  ESS_tail  ESS_bulk/s  R_hat

lp__              -19  0.069     1.4  1.2e+00 -2.2e+01 -1.9e+01    -17       317       717         818    1.0
accept_stat__    0.65  0.060    0.34     0.30  1.1e-03     0.77   1.00        43       279         110    1.1
stepsize__      0.060    nan   0.011    0.010  4.3e-02    0.063  0.071       nan       nan         nan    nan
treedepth__       3.9   0.23     1.5      1.5  1.0e+00      4.0    6.0        48        57         123    1.1
n_leapfrog__       30    5.4      26       28  2.0e+00       26     67        43       346         111    1.1
divergent__      0.36    nan    0.48     0.00  0.0e+00     0.00    1.0       214       214         550    1.0
energy__           21  0.079     1.9      1.8  1.8e+01       21     25       590      1548        1520    1.0

x[1]             0.27  0.025    0.42  1.5e-05  1.7e-23  9.9e-06   1.00       131       528         338    1.0
x[2]             0.12  0.026    0.32  1.8e-21  4.3e-72  1.2e-21   1.00       164       328         423    1.0
x[3]             0.33  0.045    0.46  5.6e-11  1.6e-63  3.8e-11    1.0       135        89         348    1.0
x[4]            0.068  0.031    0.24  5.2e-31  1.2e-79  3.5e-31   0.99       112       346         288    1.0
x[5]             0.21  0.025    0.38  3.2e-06  2.2e-16  2.2e-06   1.00        71       970         182    1.1

$ ./bin/diagnose sb_1.csv sb_2.csv sb_3.csv sb_4.csv

Checking sampler transitions treedepth.
Treedepth satisfactory for all transitions.

Checking sampler transitions for divergences.
1548 of 1000 (154.80%) transitions ended with a divergence.
These divergent transitions indicate that HMC is not fully able to explore the posterior distribution.
Try increasing adapt delta closer to 1.
If this doesn't remove all divergences, try to reparameterize the model.

Checking E-BFMI - sampler transitions HMC potential energy.
E-BFMI satisfactory.

Rank-normalized split effective sample size satisfactory for all parameters.

The following parameters had rank-normalized split R-hat greater than 1.01:
  x[1], x[2], x[3], x[5]
Such high values indicate incomplete mixing and biased estimation.
You should consider regularizating your model with additional prior information or a more effective parameterization.
```

And this PR (`ilr`):
```
$ ./bin/stansummary ilr_1.csv ilr_2.csv ilr_3.csv ilr_4.csv

Inference for Stan model: multi_normal_logit_model
4 chains: each with iter=1000; warmup=1000; thin=1; 1000 iterations saved.

Warmup took (0.20, 0.24, 0.20, 0.20) seconds, 0.83 seconds total
Sampling took (0.22, 0.19, 0.20, 0.22) seconds, 0.84 seconds total

                 Mean     MCSE   StdDev      MAD       5%      50%    95%  ESS_bulk  ESS_tail  ESS_bulk/s  R_hat

lp__              -18  3.4e-02      1.4  1.2e+00 -2.1e+01 -1.8e+01    -17      1734      2095        2071    1.0
accept_stat__    0.93    0.010  1.1e-01  4.3e-02     0.70     0.97   1.00       768      2447         917    1.0
stepsize__      0.033      nan  3.3e-03  4.8e-03    0.029    0.033  0.036       nan       nan         nan    nan
treedepth__       5.9    0.023  1.3e+00  1.5e+00      2.0      6.0    7.0      2952       695        3527    1.0
n_leapfrog__       95      1.3  4.9e+01  4.2e+01      7.0      127    127      1972      2421        2356    1.0
divergent__      0.00      nan  0.0e+00  0.0e+00     0.00     0.00   0.00       nan       nan         nan    nan
energy__           20    0.049  2.0e+00  1.8e+00       18       20     24      1696      2159        2026    1.0

x[1]            0.087  5.6e-03     0.27  1.2e-27  6.0e-75  8.1e-28   1.00      2005      2372        2396   1.00
x[2]             0.30  9.9e-03     0.46  4.4e-20  5.9e-97  3.0e-20    1.0      1806      1998        2157    1.0
x[3]             0.32  9.9e-03     0.46  8.9e-20 6.4e-103  6.0e-20    1.0      1587      2230        1896    1.0
x[4]             0.21  8.7e-03     0.40  1.3e-26  9.9e-84  8.8e-27    1.0      1798      2068        2148    1.0
x[5]            0.080  5.5e-03     0.26  2.1e-27  3.6e-75  1.4e-27   0.98      1599      2009        1910    1.0

$ ./bin/diagnose ilr_1.csv ilr_2.csv ilr_3.csv ilr_4.csv

Checking sampler transitions treedepth.
Treedepth satisfactory for all transitions.

Checking sampler transitions for divergences.
No divergent transitions found.

Checking E-BFMI - sampler transitions HMC potential energy.
E-BFMI satisfactory.

Rank-normalized split effective sample size satisfactory for all parameters.

Rank-normalized split R-hat values satisfactory for all parameters.
```

File: stan/math/prim/constraint/simplex_constrain.hpp [hunk 1]
```diff
@@ -27,29 +32,56 @@ namespace math {
...
+    max_val_old = max_val;
+  }
+
+  // above loop doesn't reach i==0
+  max_val = fmax(max_val_old, z.coeff(0));
+  d = d * exp(max_val_old - max_val) + exp(z.coeff(0) - max_val);
+
+  for (int i = 0; i <= N; ++i) {
+    z.coeffRef(i) = exp(z.coeff(i) - max_val) / d;
+  }
```
Comment (SteveBronder) [inline stan/math/prim/constraint/simplex_constrain.hpp]:
```suggestion
   z.array() = (z.array() - max_val).exp() / d;
```

**Comment by spinkney** [comment]
@WardBrian there's a bug in the logic for the sampler divergence output calc. It says 154% above

```
Checking sampler transitions for divergences.
1548 of 1000 (154.80%) transitions ended with a divergence.
```

File: stan/math/rev/constraint/stochastic_column_constrain.hpp [hunk 2]
```diff
@@ -26,53 +27,45 @@ namespace math {
...
-      stick_len_adj += arena_x_arr.row(k).adj() * arena_z_arr.row(k);
-      auto arena_z_adj = arena_x_arr.row(k).adj() * stick_len_val;
-      arena_y_arr.row(k).adj()
-          += arena_z_adj * arena_z_arr.row(k) * (1.0 - arena_z_arr.row(k));
+
+  reverse_pass_callback([arena_y, arena_x]() mutable {
+    const auto M = arena_y.cols();
+
+    const auto& x_val = to_ref(arena_x.val_op());
+    const auto& x_adj = to_ref(arena_x.adj_op());
```
Comment (SteveBronder) [inline stan/math/rev/constraint/stochastic_column_constrain.hpp]: `arena_x` should not be an expression so you don't need `to_ref` here
```suggestion
    auto&& x_val = arena_x.val_op();
    auto&& x_adj = arena_x.adj_op();
```

**Comment by WardBrian** [comment]
Yeah, it looks like the denominator never changes as you add chains. I’ll take a look up in the cmdstan repo

File: stan/math/rev/constraint/stochastic_column_constrain.hpp [hunk 3]
```diff
@@ -26,53 +27,45 @@ namespace math {
...
+
+    for (Eigen::Index i = 0; i < M; ++i) {
+      // backprop for softmax
+      Eigen::VectorXd x_pre_softmax_adj
+          = -x_val.col(i) * x_adj.col(i).dot(x_val.col(i))
+            + x_val.col(i).cwiseProduct(x_adj.col(i));
```
Comment (SteveBronder) [inline stan/math/rev/constraint/stochastic_column_constrain.hpp]: You can just make the memory `x_pre_softmax_adj` once
```suggestion
    Eigen::VectorXd x_pre_softmax_adj(x_val.rows());
    for (Eigen::Index i = 0; i < M; ++i) {
      // backprop for softmax
      x_pre_softmax_adj.noalias()
          = -x_val.col(i) * x_adj.col(i).dot(x_val.col(i))
            + x_val.col(i).cwiseProduct(x_adj.col(i));
```

Comment (SteveBronder) [reply stan/math/rev/constraint/stochastic_column_constrain.hpp]: You could also change `sum_to_zero_vector_backprop` to just take in a matrix

File: stan/math/rev/constraint/sum_to_zero_constrain.hpp [hunk 4]
```diff
@@ -15,6 +15,35 @@
...
+namespace internal {
+
+/**
+ * The reverse pass backprop for the sum_to_zero_constrain on
+ * vectors. This is separated out so it can also be called by
+ * simplex_constrain.
+ */
+template <typename T>
+void sum_to_zero_vector_backprop(T&& y_adj, const Eigen::VectorXd& z_adj) {
+  const auto N = y_adj.size();
```
Comment (SteveBronder) [inline stan/math/rev/constraint/sum_to_zero_constrain.hpp:29]: I'd leave a small comment so people know `z_adj` is `y_adj.size() + 1`

**Review by SteveBronder — DISMISSED** [review]
Good with a few small suggestions

File: stan/math/rev/constraint/sum_to_zero_constrain.hpp (hunk 4 above)
Comment (WardBrian) [reply stan/math/rev/constraint/sum_to_zero_constrain.hpp:29]: I added `@param`s to the docstring and put the sizes there

File: stan/math/rev/constraint/simplex_constrain.hpp [hunk 5]
```diff
@@ -31,38 +32,24 @@ template <typename T, require_rev_col_vector_t<T>* = nullptr>
...
-    for (Eigen::Index k = N; k-- > 0;) {
-      arena_x.adj().coeffRef(k) -= stick_len_adj;
-      stick_len_val += arena_x.val().coeff(k);
-      stick_len_adj += arena_x.adj().coeff(k) * arena_z.coeff(k);
-      double arena_z_adj = arena_x.adj().coeff(k) * stick_len_val;
-      arena_y.adj().coeffRef(k)
-          += arena_z_adj * arena_z.coeff(k) * (1.0 - arena_z.coeff(k));
-    }
+  reverse_pass_callback([arena_y, arena_x]() mutable {
+    const auto& res_val = to_ref(arena_x.val());
```
Comment (SteveBronder) [inline stan/math/rev/constraint/simplex_constrain.hpp]: You should not need `to_ref` on any of the `arena_x` vals

File: stan/math/rev/constraint/simplex_constrain.hpp [hunk 6]
```diff
@@ -73,59 +60,42 @@ inline auto simplex_constrain(const T& y) {
...
-      stick_len_adj += arena_x.adj().coeff(k) * arena_z.coeff(k);
-      double arena_z_adj = arena_x.adj().coeff(k) * stick_len_val;
-      arena_y.adj().coeffRef(k)
-          += arena_z_adj * arena_z.coeff(k) * (1.0 - arena_z.coeff(k));
-    }
+  reverse_pass_callback([arena_y, arena_x, lp]() mutable {
+    const auto& res_val = to_ref(arena_x.val());
+
+    // backprop for log jacobian contribution to log density
+    arena_x.adj().array() += lp.adj() / res_val.array();
```
Comment (SteveBronder) [inline stan/math/rev/constraint/simplex_constrain.hpp]: ~This is kind of confusing, why are we accumulating into the return values adjoints here?~
EDIT: nvm I see this is the rhs of the jacobian adjustment. Though I can't think of anywhere else we change the outputs adjoints in the reverse pass of it's function. I need to think about this

Comment (WardBrian) [reply stan/math/rev/constraint/simplex_constrain.hpp]: We can avoid it by making a copy of the adjoints if we want, though I'm pretty sure that nothing else in the AD graph can depend on the adjoints of this by this point

Comment (SteveBronder) [reply stan/math/rev/constraint/simplex_constrain.hpp]: Can you just use `auto lp_adj =` so then it's just an expression and you don't need any memory

Comment (WardBrian) [reply stan/math/rev/constraint/simplex_constrain.hpp]: Can you elaborate?

Comment (SteveBronder) [reply stan/math/rev/constraint/simplex_constrain.hpp]: I mean 
```
 auto lp_adj = lp.adj() / res_val.array();
```

then in the below you can do `res_val * (arena_x.adj() + lp_adj).dot(rev_val)`

Comment (WardBrian) [reply stan/math/rev/constraint/simplex_constrain.hpp]: `arena_x.adj()` is used two places, so you'd be duplicating the work, no?

**Comment by spinkney** [comment]
Is this waiting for my review? I can do it next week, otherwise what else is needed?

**Comment by WardBrian** [comment]
@spinkney this is waiting on final review from @SteveBronder and then us deciding if we’re okay with not preserving the existing code, otherwise it should be good

**Comment by spinkney** [comment]
Ok, I still think it's worth another issue on saving different transforms and creating how a user would specify that. For this PR I vote that we can just remove the existing code.

File: stan/math/rev/constraint/simplex_constrain.hpp (hunk 6 above)
Comment (WardBrian) [reply stan/math/rev/constraint/simplex_constrain.hpp]: I believe this is resolved, the code no longer needs to modify the adjoint of the return

**Review by SteveBronder — APPROVED** [review]
